app_test.go

```diff
@@ -2183,3 +2183,63 @@ func TestSetupInitializesOnlyNilWriters(t *testing.T) {
 		t.Errorf("expected a.Writer to be os.Stdout")
 	}
 }
+
+func TestApp_PreventDefaultPrint_When_BoolFlag(t *testing.T) {
+
+	app := App{
+		Name:  "greet",
+		Usage: "fight the loneliness!",
+		Flags: []Flag{
+			&BoolFlag{
+				Name:             "verbose",
+				Aliases:          []string{"v"},
+				Usage:            "Increase verbosity",
+				ShowDefaultValue: true,
+				DefaultText:      "default text",
+			},
+			&BoolFlag{
+				Name:             "extract",
+				Aliases:          []string{"e"},
+				Usage:            "Extract only, do not execute the service",
+				ShowDefaultValue: false,
+			},
+			&BoolFlag{
+				Name:    "keep",
+				Aliases: []string{"k"},
+				Usage:   " Do not delete the temp dir after execution",
+			},
+			&StringFlag{
+				Name:  "lang",
+				Value: "english",
+				Usage: "language for the greeting",
+			},
+		},
+
+		Action: func(c *Context) error {
+			return nil
+		},
+	}
+
+	err := app.Run([]string{"", "-h"})
+	if err != nil {
+		t.Error(err)
+	}
+	/*
```

Comment:
A test for this behavior needs to assert something other than having no errors during the run. Having this example is great, but what we really need is for that example to be proven and enforced programmatically.

In particular, if this is a change to help documentation, we should capture the output of the application (in a `bytes.Buffer` is probably easiest), then assert that the output looks like what's documented here.

Reply:
@rliebz I have used `bytes.Buffer` but i could not get proper output as expected. Actually i can not find out why `bytes.Buffer` can not convert multi line string.

would you please review the test code 
```
func TestApp_PreventDefaultPrint_When_BoolFlag(t *testing.T) {
	var actualOutput bytes.Buffer
	var expectedOutput bytes.Buffer

	expectedOutput.WriteString(
		`NAME: 
		greet - fight the loneliness!

		USAGE:
		debug.test [global options] command [command options] [arguments...]

		COMMANDS:
		help, h  Shows a list of commands or help for one command

		GLOBAL OPTIONS:
		--verbose, -v  Increase verbosity (default: default text)
		--extract, -e  Extract only, do not execute the service
		--keep, -k     Do not delete the temp dir after execution
		--help, -h     show help
		`)

	app := App{
		Name:   "greet",
		Usage:  "fight the loneliness!",
		Writer: &actualOutput,
		Flags: []Flag{
			&BoolFlag{
				Name:             "verbose",
				Aliases:          []string{"v"},
				Usage:            "Increase verbosity",
				ShowDefaultValue: true,
				DefaultText:      "default text",
			},
			&BoolFlag{
				Name:             "extract",
				Aliases:          []string{"e"},
				Usage:            "Extract only, do not execute the service",
				ShowDefaultValue: false,
			},
			&BoolFlag{
				Name:    "keep",
				Aliases: []string{"k"},
				Usage:   " Do not delete the temp dir after execution",
			},
		},

		Action: func(c *Context) error {
			return nil
		},
	}

	_ = app.Run([]string{"", "-h"})

	actualOutputString := actualOutput.String()
	expectedOutputString := expectedOutput.String()
	expect(t, actualOutputString, expectedOutputString)
}
```

Reply:
`bytes.Buffer` doesn't do anything special with newlines, but Go's multi-line strings using the backtick character might not behave as you expect. Leading whitespace is preserved, so for multi-line strings, you'll have to outdent if you don't expect leading indentation.

Something like this:
```go
		expectedOutput.WriteString(
				`NAME: 
greet - fight the loneliness!

USAGE:
...
```

It definitely looks weird, but that's the way to represent them. Unfortunately our `expect` helper function doesn't do a great job of showing you what the difference is, but if you continue to run into issues, I'd recommend looking at a tool like https://github.com/google/go-cmp to get a good line-by-line comparison as you debug.

Reply:
Hi @rliebz, I used `strings.Fields(actualOutput.String())` to compare the multi-line strings which is actually great. Could you please review?

Reply:
`strings.Fields` is not the best comparison for multi-line strings because it effectively ignores whitespace. 

If you're looking for a line-by-line breakdown, you could use `strings.Split(actualOutput.String(), "\n")`.

Also, we should remove the comment, since we have an actual test string that's being compared.